src/merge_stop_areas.rs

```diff
@@ -0,0 +1,346 @@
+            }
+            let master = stops_with_prio.remove(0);
+            let others = stops_with_prio.into_iter().map(|stop_with_prio| stop_with_prio.0).collect();
+            Some((
+                k.clone(),
+                StopAreaGroupRule {
+                    master_stop_area_id: master.0,
+                    to_merge_stop_area_ids: others,
+                },
+            ))
+        }).collect();
+    group_rules.values().into_iter().cloned().collect()
+}
+
+fn read_rules<P: AsRef<path::Path>>(
+    paths: Vec<P>,
+    report: &mut MergeStopAreasReport,
+) -> Vec<StopAreaGroupRule> {
+    let mut rules: Vec<StopAreaGroupRule> = vec![];
+    for rule_path in paths {
+        let rule_path = rule_path.as_ref();
+        let mut rdr = csv::Reader::from_path(&rule_path)
+            .with_context(ctx_from_path!(rule_path))
+            .unwrap();
+        let file_rules: Vec<StopAreaMergeRule> = rdr
+            .deserialize()
+            .collect::<StdResult<_, _>>()
+            .with_context(ctx_from_path!(rule_path))
+            .unwrap();
+        rules.extend(group_rules_from_file_rules(file_rules, report));
+    }
+    rules
+}
+
+fn generate_automatic_rules(
+    stop_areas: &CollectionWithId<StopArea>,
+    distance: u16,
+) -> Vec<StopAreaGroupRule> {
+    let mut stop_area_iter = stop_areas.values();
+    let sq_max_distance: f64 = (distance * distance).into();
+    let mut automatic_rules = vec![];
+    while let Some(top_level_stop_area) = stop_area_iter.next() {
+        for bottom_level_stop_area in stop_area_iter.clone() {
+            if top_level_stop_area.name == bottom_level_stop_area.name {
+                let approx = top_level_stop_area.coord.approx();
+                let sq_distance = approx.sq_distance_to(&bottom_level_stop_area.coord);
+                if sq_distance <= sq_max_distance {
+                    automatic_rules.push(StopAreaGroupRule {
+                        master_stop_area_id: top_level_stop_area.id.clone(),
+                        to_merge_stop_area_ids: vec![bottom_level_stop_area.id.clone()],
+                    });
+                }
+            }
+        }
+    }
+    automatic_rules
+}
+
+fn apply_rules(
+    mut collections: Collections,
+    rules: Vec<StopAreaGroupRule>,
+    report: &mut MergeStopAreasReport,
+) -> Collections {
+    let mut stop_points_updated = collections.stop_points.take();
+    let mut geometries_updated = collections.geometries.take();
+    let mut lines_updated = collections.lines.take();
+    let mut stop_areas_to_remove: Vec<String> = Vec::new();
+    let mut stop_area_ids = collections
+        .stop_areas
+        .values()
+        .map(|stop_area| stop_area.id.clone())
+        .collect::<Vec<String>>();
+    for mut rule in rules {
+        stop_area_ids.retain(|id| !stop_areas_to_remove.contains(&id));
+        rule = skip_fail!(rule.ensure_rule_valid(&stop_area_ids, report));
+        for stop_point in stop_points_updated.iter_mut() {
```

Comment:
`for stop_point in &mut stop_points_updated` should work too

Reply:
is it better?

Reply:
I don't know? @TeXitoi ?

Reply:
It's more rustic, yes.